rt_gene/src/rt_gene/estimate_gaze_pytorch.py

```diff
@@ -0,0 +1,54 @@
+# Licensed under Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International (https://creativecommons.org/licenses/by-nc-sa/4.0/legalcode)
+
+import os
+
+import cv2
+import torch
+from torchvision import transforms
+from tqdm import tqdm
+
+from rt_gene.estimate_gaze_base import GazeEstimatorBase
+from rt_gene.gaze_estimation_models_pytorch import GazeEstimationmodelResnet18
+
+
+class GazeEstimator(GazeEstimatorBase):
+    """This class encapsulates a deep neural network for gaze estimation.
+
+    It retrieves two image streams, one containing the left eye and another containing the right eye.
+    It synchronizes these two images with the estimated head pose.
+    The images are then converted in a suitable format, and a forward pass of the deep neural network
+    results in the estimated gaze for this frame. The estimated gaze is then published in the (theta, phi) notation."""
+
+    def __init__(self, device_id_gaze, model_files):
+        super(GazeEstimator, self).__init__(device_id_gaze, model_files)
+        if "OMP_NUM_THREADS" not in os.environ:
+            os.environ["OMP_NUM_THREADS"] = "8"
+        tqdm.write("PyTorch using {} threads.".format(os.environ["OMP_NUM_THREADS"]))
+
+        self._transform = transforms.Compose([lambda x: cv2.resize(x, dsize=(224, 224), interpolation=cv2.INTER_CUBIC),
+                                              transforms.ToTensor(),
+                                              transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])])
+
+        self._models = []
+        for ckpt in self.model_files:
+            _model = GazeEstimationmodelResnet18(num_out=2)
+            _torch_load = torch.load(ckpt)['state_dict']
+            _state_dict = {k[7:]: v for k, v in _torch_load.items()}
```

Comment:
Can we comment this magic line somehow?

Reply:
agreed; I've commented it but it's to do with the way pytorch_lightning saves checkpoints.
The pl.LightningModule is actually a nn.Module itself and thus saves it's children. It's children include the criterion and the model (as _criterion and _model).

When it comes to loading the checkpoint, we technically should be loading the pl.LightningModule but that would load it's criterion and overhead. Instead, I'm only really after loading the _model itself. Unfortunately the state_dict appends _model in-front of the layer names and thus just removes _model from it; thus avoiding loading the pl.LightningModule and the _criterion.

Reply:
done.